include/tscore/MT_hashtable.h

```diff
@@ -212,20 +223,21 @@ IMTHashTable<key_t, data_t>::insert_entry(key_t key, data_t data)
 {
   int id                             = bucket_id(key);
   HashTableEntry<key_t, data_t> *cur = buckets[id];
-  while (cur != NULL && cur->key != key) {
-    cur = cur->next;
-  }
-  if (cur != NULL) {
-    if (data == cur->data) {
-      return (data_t)0;
-    } else {
-      data_t tmp = cur->data;
-      cur->data  = data;
-      // potential memory leak, need to check the return value by the caller
-      return tmp;
+  if (m_key_unique) {
+    while (cur != NULL && cur->key != key) {
+      cur = cur->next;
+    }
+    if (cur != NULL) {
+      if (data == cur->data) {
+        return (data_t)0;
```

Comment:
No C-style casts! This should be `nullptr`, which removes the need for casting. Otherwise, you can't presume 0 can be cast to `data_t`.